src/engine.ml

```diff
@@ -906,36 +958,72 @@ let server_send_push_reply config is_not_taken tls session key tls_data =
   in
   Ok (Some (`Established ip_config), config', channel_st, [ (`Control, out) ])
 
-let server_handle_tls_data config is_not_taken session keys tls d =
+let server_handle_tls_data config auth_user_pass is_not_taken session keys tls d
+    =
   let open Result.Syntax in
-  let* next, config, out = kex_server config session keys tls d in
+  let* next, config, out =
+    kex_server config auth_user_pass session keys tls d
+  in
+  let out = List.map (fun out -> (`Control, out)) out in
   match next with
   | `Send_push_reply (tls', key, tls_data) ->
       let* ip_config, config, channel_st, out' =
         server_send_push_reply config is_not_taken tls' session key tls_data
       in
-      Ok (ip_config, config, channel_st, (`Control, out) :: out')
+      Ok (ip_config, config, channel_st, out @ out')
   | `State (channel_st, ip_config) ->
       (* keys established, move forward to "expect push request (reply with push reply)" *)
-      Ok (ip_config, config, channel_st, [ (`Control, out) ])
+      Ok (ip_config, config, channel_st, out)
+  | `Authentication_failed _tls -> Ok (Some `Exit, config, Expect_reset, out)
 
-let incoming_control_server is_not_taken config rng session channel _now _ts
-    _key op data =
+let incoming_control_server auth_user_pass is_not_taken config rng session
+    channel now _ts _key op data =
+  let open Result.Syntax in
   match (channel.channel_st, op) with
   | ( Expect_reset,
       ( Packet.Hard_reset_client_v2 | Packet.Hard_reset_client_v3
       | Packet.Soft_reset_v2 ) ) ->
-      (* TODO may need to do client certificate authentication here! *)
-      let _ca, server, key =
-        ( Config.get Ca config,
-          Config.get Tls_cert config,
-          Config.get Tls_key config )
+      let server, key =
+        (Config.get Tls_cert config, Config.get Tls_key config)
       in
       let ciphers = tls_ciphers config and version = tls_version config in
+      let* authenticator =
+        match
+          ( Config.find Verify_client_cert config,
```

Comment:
Can be done in a different commit, but `verify-client-cert required` is the default according to the man page.

Reply:
It does not make a big difference whether we synthesize here or in config.ml...

Reply:
yes, well, there's the pain of matching on ``None | Some `Required`` -- but I guess a refactoring may be to move from config to some custom record where only the stuff really needed is present... at the cost of having yet another record type

or to fill the verify-client-cert in valid_server_config if not provided, and use here a `Config.get` instead of `Config.find`.